File: reference-impl/typescript/aad-builder.test.ts
import { buildAad, hexToBytes, bytesToHex, uuidToNetworkBytes } from './aad-builder';
import * as fs from 'fs';
import * as path from 'path';

interface TestVector {
  id: string;
  description: string;
  inputs: {
    aeadSuiteId: number;
    hashSuiteId: number;
    evidenceId: string;
    chunkIndex: number;
    manifestVer: number;
    policyHash: string;
  };
  expected: {
    aadHex: string;
    aadLength: number;
  };
}

interface TestVectorFile {
  vectors: TestVector[];
}

describe('AAD Builder', () => {
  let testVectors: TestVector[];

  beforeAll(() => {
    const vectorsPath = path.join(__dirname, '../../test-vectors/aad-computation.json');
    const data: TestVectorFile = JSON.parse(fs.readFileSync(vectorsPath, 'utf-8'));
    testVectors = data.vectors;
  });

  describe('UUID to network bytes', () => {
    it('should convert UUID to big-endian bytes', () => {
      const uuid = '550e8400-e29b-41d4-a716-446655440000';
      const bytes = uuidToNetworkBytes(uuid);
      const hex = bytesToHex(bytes);
      expect(hex).toBe('550e8400e29b41d4a716446655440000');
    });

    it('should handle all-zeros UUID', () => {
      const uuid = '00000000-0000-0000-0000-000000000000';
      const bytes = uuidToNetworkBytes(uuid);
      expect(bytes.length).toBe(16);
      expect(bytes.every(b => b === 0)).toBe(true);
    });
  });

  describe('AAD computation', () => {
    it('should produce correct AAD for test vector aad-001', () => {
      const vector = testVectors.find(v => v.id === 'aad-001');
      if (!vector) throw new Error('Vector aad-001 not found');

      const aad = buildAad({
        aeadSuiteId: vector.inputs.aeadSuiteId,
        hashSuiteId: vector.inputs.hashSuiteId,
        evidenceId: vector.inputs.evidenceId,
        chunkIndex: BigInt(vector.inputs.chunkIndex),
        manifestVer: vector.inputs.manifestVer,
        policyHash: hexToBytes(vector.inputs.policyHash),
      });

      expect(aad.length).toBe(60);
      expect(bytesToHex(aad)).toBe(vector.expected.aadHex.toLowerCase());
    });

    it('should produce correct AAD for test vector aad-002 (second chunk)', () => {
      const vector = testVectors.find(v => v.id === 'aad-002');
      if (!vector) throw new Error('Vector aad-002 not found');

      const aad = buildAad({
        aeadSuiteId: vector.inputs.aeadSuiteId,
        hashSuiteId: vector.inputs.hashSuiteId,
        evidenceId: vector.inputs.evidenceId,
        chunkIndex: BigInt(vector.inputs.chunkIndex),
        manifestVer: vector.inputs.manifestVer,
        policyHash: hexToBytes(vector.inputs.policyHash),
      });

      expect(bytesToHex(aad)).toBe(vector.expected.aadHex.toLowerCase());
    });

    it('should produce correct AAD for test vector aad-003 (large chunk index)', () => {
      const vector = testVectors.find(v => v.id === 'aad-003');
      if (!vector) throw new Error('Vector aad-003 not found');

      const aad = buildAad({
        aeadSuiteId: vector.inputs.aeadSuiteId,
        hashSuiteId: vector.inputs.hashSuiteId,
        evidenceId: vector.inputs.evidenceId,
        chunkIndex: BigInt(vector.inputs.chunkIndex),
        manifestVer: vector.inputs.manifestVer,
        policyHash: hexToBytes(vector.inputs.policyHash),
      });

      expect(bytesToHex(aad)).toBe(vector.expected.aadHex.toLowerCase());
    });

    it('should handle AES-256-GCM-SIV suite (aad-004)', () => {
      const vector = testVectors.find(v => v.id === 'aad-004');
      if (!vector) throw new Error('Vector aad-004 not found');

      const aad = buildAad({
        aeadSuiteId: vector.inputs.aeadSuiteId,
        hashSuiteId: vector.inputs.hashSuiteId,
        evidenceId: vector.inputs.evidenceId,
        chunkIndex: BigInt(vector.inputs.chunkIndex),
        manifestVer: vector.inputs.manifestVer,
        policyHash: hexToBytes(vector.inputs.policyHash),
      });

      expect(aad[0]).toBe(0x02); // AES-256-GCM-SIV
      expect(bytesToHex(aad)).toBe(vector.expected.aadHex.toLowerCase());
    });

    it('should validate policyHash length', () => {
      expect(() => {
        buildAad({
          evidenceId: '550e8400-e29b-41d4-a716-446655440000',
          chunkIndex: 0n,
          policyHash: new Uint8Array(31), // Wrong length
        });
      }).toThrow('policyHash must be 32 bytes');
    });

    it('should reject negative chunk index', () => {
      expect(() => {
        buildAad({
          evidenceId: '550e8400-e29b-41d4-a716-446655440000',
          chunkIndex: -1n,
          policyHash: new Uint8Array(32),
        });
      }).toThrow('chunkIndex must be non-negative');
    });
  });

  describe('All test vectors', () => {
    it('should pass all AAD test vectors', () => {
      for (const vector of testVectors) {
        const aad = buildAad({
          aeadSuiteId: vector.inputs.aeadSuiteId,
          hashSuiteId: vector.inputs.hashSuiteId,
          evidenceId: vector.inputs.evidenceId,
          chunkIndex: BigInt(vector.inputs.chunkIndex),
          manifestVer: vector.inputs.manifestVer,
          policyHash: hexToBytes(vector.inputs.policyHash),
        });

        const computed = bytesToHex(aad);
        const expected = vector.expected.aadHex.toLowerCase();

        expect(computed).toBe(expected);
      }
    });
  });
});

File: reference-impl/kotlin/test/AadBuilderTest.kt
package com.evidence.contracts

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.TestInstance
import java.io.File
import java.util.UUID
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class AadBuilderTest {

    data class TestVector(
        val id: String,
        val description: String,
        val inputs: Inputs,
        val expected: Expected
    )

    data class Inputs(
        val aeadSuiteId: Int,
        val hashSuiteId: Int,
        val evidenceId: String,
        val chunkIndex: Long,
        val manifestVer: Int,
        val policyHash: String
    )

    data class Expected(
        val aadHex: String,
        val aadLength: Int
    )

    data class TestVectorFile(
        val vectors: List<TestVector>
    )

    private lateinit var testVectors: List<TestVector>

    @BeforeAll
    fun loadTestVectors() {
        val vectorsFile = File("../../test-vectors/aad-computation.json")
        val json = vectorsFile.readText()
        val type = object : TypeToken<TestVectorFile>() {}.type
        val data: TestVectorFile = Gson().fromJson(json, type)
        testVectors = data.vectors
    }

    @Test
    fun `UUID to network bytes should be big-endian`() {
        val uuid = UUID.fromString("550e8400-e29b-41d4-a716-446655440000")
        val bytes = AadBuilder.uuidToNetworkBytes(uuid)
        val hex = bytes.toHex()
        assertEquals("550e8400e29b41d4a716446655440000", hex)
    }

    @Test
    fun `AAD for test vector aad-001`() {
        val vector = testVectors.find { it.id == "aad-001" }
            ?: throw AssertionError("Vector aad-001 not found")

        val aad = AadBuilder.build(
            aeadSuiteId = vector.inputs.aeadSuiteId.toByte(),
            hashSuiteId = vector.inputs.hashSuiteId.toByte(),
            evidenceId = UUID.fromString(vector.inputs.evidenceId),
            chunkIndex = vector.inputs.chunkIndex,
            manifestVer = vector.inputs.manifestVer,
            policyHash = vector.inputs.policyHash.hexToBytes()
        )

        assertEquals(60, aad.size)
        assertEquals(vector.expected.aadHex.lowercase(), aad.toHex())
    }

    @Test
    fun `AAD for test vector aad-002 (second chunk)`() {
        val vector = testVectors.find { it.id == "aad-002" }
            ?: throw AssertionError("Vector aad-002 not found")

        val aad = AadBuilder.build(
            aeadSuiteId = vector.inputs.aeadSuiteId.toByte(),
            hashSuiteId = vector.inputs.hashSuiteId.toByte(),
            evidenceId = UUID.fromString(vector.inputs.evidenceId),
            chunkIndex = vector.inputs.chunkIndex,
            manifestVer = vector.inputs.manifestVer,
            policyHash = vector.inputs.policyHash.hexToBytes()
        )

        assertEquals(vector.expected.aadHex.lowercase(), aad.toHex())
    }

    @Test
    fun `AAD for test vector aad-003 (large chunk index)`() {
        val vector = testVectors.find { it.id == "aad-003" }
            ?: throw AssertionError("Vector aad-003 not found")

        val aad = AadBuilder.build(
            aeadSuiteId = vector.inputs.aeadSuiteId.toByte(),
            hashSuiteId = vector.inputs.hashSuiteId.toByte(),
            evidenceId = UUID.fromString(vector.inputs.evidenceId),
            chunkIndex = vector.inputs.chunkIndex,
            manifestVer = vector.inputs.manifestVer,
            policyHash = vector.inputs.policyHash.hexToBytes()
        )

        assertEquals(vector.expected.aadHex.lowercase(), aad.toHex())
    }

    @Test
    fun `AAD should handle AES-256-GCM-SIV suite`() {
        val vector = testVectors.find { it.id == "aad-004" }
            ?: throw AssertionError("Vector aad-004 not found")

        val aad = AadBuilder.build(
            aeadSuiteId = vector.inputs.aeadSuiteId.toByte(),
            hashSuiteId = vector.inputs.hashSuiteId.toByte(),
            evidenceId = UUID.fromString(vector.inputs.evidenceId),
            chunkIndex = vector.inputs.chunkIndex,
            manifestVer = vector.inputs.manifestVer,
            policyHash = vector.inputs.policyHash.hexToBytes()
        )

        assertEquals(0x02, aad[0].toInt()) // AES-256-GCM-SIV
        assertEquals(vector.expected.aadHex.lowercase(), aad.toHex())
    }

    @Test
    fun `AAD should validate policyHash length`() {
        assertFailsWith<IllegalArgumentException> {
            AadBuilder.build(
                evidenceId = UUID.randomUUID(),
                chunkIndex = 0,
                policyHash = ByteArray(31) // Wrong length
            )
        }
    }

    @Test
    fun `AAD should reject negative chunk index`() {
        assertFailsWith<IllegalArgumentException> {
            AadBuilder.build(
                evidenceId = UUID.randomUUID(),
                chunkIndex = -1,
                policyHash = ByteArray(32)
            )
        }
    }

    @Test
    fun `All test vectors should pass`() {
        for (vector in testVectors) {
            val aad = AadBuilder.build(
                aeadSuiteId = vector.inputs.aeadSuiteId.toByte(),
                hashSuiteId = vector.inputs.hashSuiteId.toByte(),
                evidenceId = UUID.fromString(vector.inputs.evidenceId),
                chunkIndex = vector.inputs.chunkIndex,
                manifestVer = vector.inputs.manifestVer,
                policyHash = vector.inputs.policyHash.hexToBytes()
            )

            val computed = aad.toHex()
            val expected = vector.expected.aadHex.lowercase()

            assertEquals(expected, computed, "Failed for vector ${vector.id}")
        }
    }
}
